### DISCUSSION TITLE
NPC respawn

### DISCUSSION THREAD

**Discussion by owllex** [evennia/evennia#2449] [Evennia Questions] [answered]
Hey everyone,

Is there any sort of best practice for how to handle NPC respawns? To be clear, the behavior I'm looking for is to have certain characters reappear in their home location with full vitals and with all of their original items in some period of time after they have died.

In my game, I've defined an `NPC` typeclass that inherits from `Character` to hold NPC-specific logic. Characters can die in a number of ways, but most commonly through losing all of their HP, and I have an `at_death()` handler that gets called when this happens. Right now, `at_death()` spawns a corpse object in the current location that contains their inventory.

For player characters, my plan is to teleport the character to a special afterlife zone that provides an IC way to get a resurrection (aka moving them back into the regular chunk of the game world).

For NPCs, I just want them to disappear until some time has passed, and then reappear back in their home location. I had a few thoughts about how to deal with NPC respawns and wanted to get some advice. Beyond just dropping a corpse, here were a couple thoughts I had:

1. `at_death()` moves the NPC into an inaccessible NPC graveyard room, and then a revive script restores their vitals and moves them back after some amount of time.
* Upsides: The same `NPC` object can be continually re-used. No need to actually re-spawn it from a prototype.
* Downsides: This graveyard room could get very full if a lot of NPCs die at once. I don't know if this becomes a performance problem at some point (my game is small, but I like to think about scaling issues regardless). Also, this solution doesn't handle the NPC's inventory. I'd probably need to have the revive script re-spawn all of the NPC's inventory and provide a way to save that prototype inventory somehow.
2. `at_death()` hides the NPC, keeping it in the same room but making it super-invisible via locks to prevent it from being searchable, lookable etc. A revive script on the room then restores the NPC like in option 1.
* Upsides: Same as 1, but avoids any potentially issues with many dead NPCs hanging out in the same location.
* Downsides: Slightly goofy for admins who will be able to see all of these "ghost" NPCs everywhere since they ignore locks. (I'd probably need to hide them from look descriptions just for sanity). Still needs to handle inventory re-spawn somehow.
3. `at_death()` deletes the NPC, and then a revive script re-creates the NPC (and their inventory) directly from their prototype.
* Upsides: Straightforward. Inventory can just be built into the NPC prototype.
* Downsides: Requires a stable prototype for every respawnable NPC, which makes building a bit more complicated. Also burns through DB IDs faster (this probably isn't really a concern, but it just feels wasteful compared to option 1).

There are probably some other options too, I'd love to get some feedback about these and see if anyone else has thought about how to efficiently solve this problem.

**Comment by Griatch** [comment, marked as answer]
All of your suggestions are of course possible; let me give some input:
> 1. at_death() moves the NPC into an inaccessible NPC graveyard room, and then a revive script restores their vitals and moves them back after some amount of time.
> Downsides: This graveyard room could get very full if a lot of NPCs die at once.

There is no need for a graveyard room at all. Just set their `.location` to `None`. You can still reference the object, but it's naturally inaccessible since it's no longer in the game-world. Your revive script or timer (delay) can be used to restore them and just move them from their none-location back onto the grid. 

> Also, this solution doesn't handle the NPC's inventory. 

No, one would need to analyze what the NPC _should_ have and what it has and spawn anything that's missing.

> 2. at_death() hides the NPC, keeping it in the same room but making it super-invisible via locks to prevent it from being searchable, lookable etc. A revive script on the room then restores the NPC like in option 1.

I think this is a little more cumbersome. Just setting `location=None` is a lot easier. But I guess it could be useful if you want it to double as some sort of 'ghost' or resurrection mechanic (the dead lingers near their corpse etc). But then again, the corpse could just retain a reference to the NPC and fetch it from its `None`-location at will anyway. So yeah, probably not worth the complexity.

> 3. at_death() deletes the NPC, and then a revive script re-creates the NPC (and their inventory) directly from their prototype.

While conceptually simple, the main downside is actually performance - creating/deleting is expensive (comparably). Just changing the `location` field and changing some values back to their defaults is usually a lot cheaper computing-wise. As usual, it depends on scale. If your NPCs are few, this may be the easiest solution. It also has lesser risks of bugs (no changing things in-place and having to consider any unexpected changes, just delete and create fresh). 

> Upsides: Straightforward. Inventory can just be built into the NPC prototype.

Not sure how this would work? The contents of the inventory each have their own prototype (or could have, at least), so you'd at the very least spawn all the objects based on their prototypes and add them to the NPC (you could by all means make item prototypes with the 'location' field being the NPC off the bat, but you probably want item prototypes to be more generic than that) ...

> Downsides: Requires a stable prototype for every respawnable NPC, which makes building a bit more complicated. 

If you want the same NPC character over and over, then yeah, I suppose you need a specific prototype for it (or a custom typeclass, but that feels like overkill). 

> Also burns through DB IDs faster (this probably isn't really a concern, but it just feels wasteful compared to option `#1`).

The DB IDs are not a concern _at all_ for this (seriously, it'd be _centuries_ before you'd run out of IDs, and that's if you created 10 000 objects a second).

**Reply by owllex** [reply]
Thanks for the input. I didn't realize you could just set location to None and let them float in the void for a while. That sounds like a much better solution, thanks! I'll give it a try.

**Reply by strikaco** [reply]
> Just setting location=None is a lot easier.

I would advise against this; `location=None` used to be what determined whether an object is actually a Room. You never know what code or whose advice might still be adhering to that logic; holdouts like #2362 do exist.

Move it to Limbo instead.

**Reply by Griatch** [reply]
Nonsense. We use `location=None` for all Characters when they log out. :)

**Reply by kovitikus** [reply]
Let me just clarify here for @owllex

It doesn't matter if they are in the void or in a room. You could literally build all of your entire game in one room. The problem comes when things start interacting with each other in the room. So long as your NPCs are in a room and completely inactive, there will be no additional performance than if they were in the None void.

I set up special rooms in my own project. One is a black hole where any object that enters it will automatically be deleted. Another is a trash bin where objects stay alive in there for 90 days, after which I make sure to log them (in case they need to be restored past their expiration) and then destroy them. I have yet another room called a stasis chamber, where I send items that are to NEVER be deleted.

As long as I'm not performing consistent and extensive actions on all of these items, they won't impact the performance of my game. A room is just a container, just like any other DefaultObject. [It inherits from DefaultObject](https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L2250) and sets a few custom things on it, but it's really almost identical to any other object in Evennia.

**Reply by strikaco** [reply]
> So long as your NPCs are in a room and completely inactive, there will be no additional performance than if they were in the None void.

This is something I've been curious about and meaning to test. 

Every object/room has a ContentsHandler, which caches contents in a list, right? If you move 1M objects to Limbo, a list with 1M+ entries in it now exists in-memory (and an attempt to _render_ that abomination will presumably be made by every Character subsequently moved there). Diku had hard limits on the number of objects that another could contain (1000? 10000?).

There is no ContentsHandler for `None`. It should be _much_ more efficient to take objects off-grid, but at risk of Evennia making some annoying presumptions about the nature of objects based on arbitrary features.

**Reply by Griatch** [reply]
If you place 1M in a room, you'd better do so in a way that does not trigger the `.contents` update (so no `move_to` and maybe just setting the `db_location` field directly - and of course never actually _go_ to that room). Because yeah, if those objects ever get told to be 'aware' of each other, there will be great sorrow. Storing in `None` is indeed the way to go.

**Reply by kovitikus** [reply]
Perhaps then I'll change my strategy from my room-based sorting of objects to some sort of simple tag-based system that a script checks for every so often and decides what to do with the tagged item.

I think I can keep the black hole room, because everything that gets sent there is deleted automatically. I use it to set the default_location field of an object that I know is disposable. This way if the room it is in gets deleted and the object itself is never manually cleaned up, it will be automatically cleaned up when it is sent to the black hole.

But I suppose for other objects I plan to keep for an extended period of time, it might be better to simply tag it as immortal or indestructible or some sort of thing.

The trash bin objects already gain an attribute that sets the date they should be deleted, and it doesn't have to be exact, so I could just run a script once a day to check for all items tagged as pending deletion and see if their time has expired.

**Comment by kovitikus** [comment]
When the NPC dies, I assume that you are moving the inventory contents from the NPC object to the corpse object.

You can instead copy the objects to the corpse, retaining the original objects in the NPC's inventory.

https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L995
